cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -91,43 +196,116 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
   @Override
   public List<String> getScheduleIds(Id.Program program, SchedulableProgramType programType)
     throws SchedulerException {
-    return ImmutableList.of();
+    char startChar = ':';
+    char endChar = (char) (startChar + 1);
+    String programScheduleId = getProgramScheduleId(program, programType);
+    return ImmutableList.copyOf(scheduleSubscribers.subMap(String.format("%s%c", programScheduleId, startChar),
+                                                           String.format("%s%c", programScheduleId, endChar))
+                                  .keySet());
   }
 
   @Override
   public void suspendSchedule(Id.Program program, SchedulableProgramType programType, String scheduleName)
-    throws NotFoundException, SchedulerException {
-
+    throws ScheduleNotFoundException, SchedulerException {
+    String scheduleId = getScheduleId(program, programType, scheduleName);
+    StreamSubscriber subscriber = scheduleSubscribers.get(scheduleId);
+    if (subscriber == null) {
+      throw new ScheduleNotFoundException(scheduleName);
+    }
+    subscriber.suspendScheduleTask(program, programType, scheduleName);
   }
 
   @Override
   public void resumeSchedule(Id.Program program, SchedulableProgramType programType, String scheduleName)
-    throws NotFoundException, SchedulerException {
-
+    throws ScheduleNotFoundException, SchedulerException {
+    String scheduleId = getScheduleId(program, programType, scheduleName);
+    StreamSubscriber subscriber = scheduleSubscribers.get(scheduleId);
+    if (subscriber == null) {
+      throw new ScheduleNotFoundException(scheduleName);
+    }
+    subscriber.resumeScheduleTask(program, programType, scheduleName);
   }
 
   @Override
   public void updateSchedule(Id.Program program, SchedulableProgramType programType, Schedule schedule)
     throws NotFoundException, SchedulerException {
+    Preconditions.checkArgument(schedule instanceof StreamSizeSchedule,
+                                "Schedule should be of type StreamSizeSchedule");
+    StreamSizeSchedule streamSizeSchedule = (StreamSizeSchedule) schedule;
+    StreamSubscriber subscriber = scheduleSubscribers.get(getScheduleId(program, programType, schedule.getName()));
+    if (subscriber == null) {
+      throw new ScheduleNotFoundException(schedule.getName());
+    }
 
+    if (!streamSizeSchedule.getStreamName().equals(subscriber.getStreamId().getName())) {
+      // For a change of stream, it's okay to delete the schedule and recreate it
+      try {
+        deleteSchedule(program, programType, schedule.getName());
+      } catch (NotFoundException e) {
+        // It can happen that the schedule was deleted while being updated. In which case, the update action
+        // came first and we still want to create it
+        // TODO is it really the behavior that we want here?
```

Comment:
I don't think this can happen, because all this happens under a single transaction. the fact that the subscriber for this schedule was there means that it must be there, and under snapshot isloation, you won't see a change in the same transaction. I almost feel that this would be an internal error. For now I think it is ok to keep this code as is, but change the log level to warning, so that we see it in the logs if it happens.